Add support for opening file starting at a given line number

Lidenburg opened the pull request:
Add support for opening file starting at a given line number.

From #1185 

Some things to note: 
- Only works with less as pager. Does nothing if piped to a file.
- Files that would normally quit because of `--quit-if-one-screen` don't if an offset is provided (because of less behavior)
- If an offset that is greater than the file is long is supplied, an ugly `Cannot seek to line number N  (press RETURN)` is shown (because of less behavior)

sharkdp commented:
Thank you for your contribution!

I have a few concerns / questions:

- This would be the first feature that is really `less`-specific. If a user actually runs another pager, the `--offset` argument would be without effect. Even if `less` is the de-facto standard, I'm not sure if we should start introducing `less`-specific features. If we decide do integrate this, the limitation should be documented.
- Does this work in combination with `--line-range`? Say I run `bat --line-range=100:200 --offset 150 …`. 
- The `bat` header is not always visible. Adding a fixed delta of 3 lines is not going to work for other `--style`s. E.g.: `bat --plain --offset 50 …`.

sharkdp commented on src/bin/bat/app.rs:
```diff
@@ -238,6 +238,10 @@ impl App {
                 .map(LineRanges::from)
                 .map(|lr| HighlightedLineRanges(lr))
                 .unwrap_or_default(),
+            offset: match self.matches.value_of("offset") {
+                Some(v) => v.parse::<usize>().unwrap_or(0),
+                _ => 0,
+            }
```
You might be able to use `map_or` here:
```
offset: self.matches.value_of("offset").map_or(…)
```

Lidenburg replied:
Is something like d3e2bb3346ceaa0c4a3fcb6aab19920f1667f1c1 what you had in mind?

Lidenburg commented:
1. I have added some extra checks to only show the `--offset` argument if no pager environment variable is set or if it is set to less. It is still possible for it to have no effect if the user were to use `--pager more` in combination with `--offset` however. Not sure what would be a good way to handle this.
2. The `--offset` parameter acts as an offset from the `--line-range` starting offset, I believe this is the most intuitive behavior?
3. I think I have fixed this with commit d3e2bb3346ceaa0c4a3fcb6aab19920f1667f1c1. Unless there are times when the headers and grids can be bigger or smaller.

sharkdp commented:
>     1. I have added some extra checks to only show the `--offset` argument if no pager environment variable is set or if it is set to less.

Thank you, but I think that's kind of confusing. I'd rather always show that option and add the restriction in the `--help` text (only works with `less` as a pager).

> The `--offset` parameter acts as an offset from the `--line-range` starting offset, I believe this is the most intuitive behavior?

Sounds good.

sharkdp commented on src/bin/bat/app.rs:
```diff
@@ -223,6 +223,12 @@ impl App {
             },
+            offset: self.matches.value_of("offset").map_or(0, |v|
+                        v.parse::<usize>().unwrap()
+                        + if style_components.header() {1} else {0} // the header is 1 row
```
I think it would be better if we add a method to the style components that returns their potential header-height in lines. This way, we could use something like:
```
style_components.map(|s| s.height_in_header()).sum()
```
to get the total height.

sharkdp commented:
Closing this for now, but happy to reopen if someone wants to make progress on this.
